master/buildbot/process/build.py

```diff
@@ -284,30 +286,14 @@ def startBuild(self, build_status, workerforbuilder):
             # handler
             log.msg("Build.setupBuild failed")
             log.err(Failure())
-            self.finished = True
-            self.results = EXCEPTION
-            self.deferred = None
+            self.buildFinished(['Build.setupBuild', 'failed'], EXCEPTION)
             return
         # flush properties in the beginning of the build
         yield self._flushProperties(None)
-        yield self.master.data.updates.setBuildStateString(self.buildid,
-                                                           u'starting')
-        self.build_status.buildStarted(self)
```

Comment:
Move `build_status.buildStarted` before `Build.setupBuild`, since the later can fail triggering `Build.buildFinished` (which then calls `build_status.buildFinished`). Move the data update along with it, so they don't get out-of-sync.